tests/lib/rules/no-callback-leaks-in-ember-objects.js

```diff
@@ -0,0 +1,152 @@
+// ------------------------------------------------------------------------------
+// Requirements
+// ------------------------------------------------------------------------------
+
+const rule = require('../../../lib/rules/no-callback-leaks-in-ember-objects');
+const RuleTester = require('eslint').RuleTester;
+
+// ------------------------------------------------------------------------------
+// Tests
+// ------------------------------------------------------------------------------
+
+
+const message = "Event listeners and interval timers must be unregistered or ensure that the context they're registered with is destroyed, when no longer needed.";
+
+const eslintTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 2018,
+    sourceType: 'module'
+  }
+});
+eslintTester.run('no-callback-leaks-in-ember-objects', rule, {
+  valid: [
+    {
+      code: `
+export default Ember.Component.extend({
+  didInsertElement() {
+    if (this.get('onScroll')) {
+      this._onScrollHandler = (...args) => this.get('onScroll')(...args);
+      window.addEventListener('scroll', this._onScrollHandler);
+    }
+  },
+
+  willDestroyElement() {
+    window.removeEventListener("scroll", this._onScrollHandler);
+    this._super(...arguments);
+  }
+});`,
+    },
+    {
+      code: `
+import Component from '@ember/component';
+
+export default Component.extend({
+  didInsertElement() {
+    if (this.get('onScroll')) {
+      this._onScrollHandler = (...args) => this.get('onScroll')(...args);
+      window.addEventListener('scroll', this._onScrollHandler);
+    }
+  },
+
+  willDestroyElement() {
+    window.removeEventListener("scroll", this._onScrollHandler);
+    this._super(...arguments);
+  }
+});
+      `,
+    },
+    {
+      code: `
+import Component from '@ember/component';
+
+export default Component.extend({});`,
+    },
+    {
+      code: `
+export default Ember.Service.extend({
+  init() {
+    if (this.get('onScroll')) {
+      this._onScrollHandler = (...args) => this.get('onScroll')(...args);
+      window.addEventListener('scroll', this._onScrollHandler);
+    }
+  },
+
+  willDestroy() {
+    window.removeEventListener("scroll", this._onScrollHandler);
+    this._super(...arguments);
+  }
+});`,
+    },
+  ],
+  invalid: [
```

Comment:
I can't really tell the difference between most of these invalid test cases.

It's not worth having separate test cases just both `Ember.Component.extend` and `Component.extend`.

I would get rid of the duplicate test cases and keep only these test cases:

1. One callback which gets leaked
2. Two callbacks which get leaked
3. One callback which gets leaked and one callback which doesn't get leaked

If there are more legitimate test cases, definitely include them, but it would be helpful to add a comment explaining what each test case is testing.

Reply:
We are now having 3 invalid cases as per you mentioned, will add more if possible later